src/net_processing.cpp

```diff
@@ -3265,6 +3315,54 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
             }
         }
 
+        // Check that outbound peers have reasonable chains
+        // GetTime() is used by this anti-DoS logic so we can test this using mocktime
+        int64_t time_in_seconds = GetTime();
+        if (!state.m_chain_sync.m_protect && IsOutboundPeer(pto) && state.fSyncStarted) {
+            // This is an outbound peer subject to disconnection if they don't
+            // announce a block with as much work as the current tip within
+            // CHAIN_SYNC_TIMEOUT + HEADERS_RESPONSE_TIME seconds (note: if
+            // their chain has more work than ours, we should sync to it,
+            // unless it's invalid, in which case we should find that out and
+            // disconnect from them elsewhere).
+            if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork) {
+                if (state.m_chain_sync.m_timeout != 0) {
+                    state.m_chain_sync.m_timeout = 0;
+                    state.m_chain_sync.m_work_header = nullptr;
+                    state.m_chain_sync.m_sent_getheaders = false;
+                }
+            } else if (state.m_chain_sync.m_timeout == 0 || (state.m_chain_sync.m_work_header != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.m_chain_sync.m_work_header->nChainWork)) {
```

Comment:
I believe peers can avoid being added as disconnection candidates if they just never respond to requests for blocks/headers (pindexBestKnownBlock is unset). In that case (assuming the "Start block sync" fired off an initial GETHEADERS above), should the clock also start here if pindexBestKnownBlock is null?

Note: This comment conflicts with the one above about possibly asserting pindexBestKnownBlock here. Ignore whichever doesn't apply :)

Reply:
If they don't respond, doesn't the timeout just expire and pindexBestKnownBlock doesn't matter?

Reply:
Nice catch, thanks!

Reply:
ok, agreed.